src/sentry/static/sentry/app/views/eventsV2/saveQueryButton.tsx

```diff
@@ -0,0 +1,142 @@
+  eventView: EventView;
+  location: Location;
+  isEditing: boolean;
+};
+
+type State = {
+  queryName: string;
+};
+
+class EventsSaveQueryButton extends React.Component<Props, State> {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      queryName: props.isEditing ? props.eventView.name : '',
+    };
+  }
+
+  swallowEvent = (event: React.MouseEvent) => {
+    // Stop propagation for the input and container so
+    // people can interact with the inputs in the dropdown.
+    const capturedElements = ['LI', 'INPUT'];
+    if (
+      event.target instanceof Element &&
+      capturedElements.includes(event.target.nodeName)
+    ) {
+      event.preventDefault();
+      event.stopPropagation();
+    }
+  };
+
+  handleSave = () => {
+    const {api, eventView, organization, location, isEditing} = this.props;
+
+    const payload = eventView.toNewQuery();
+    if (this.state.queryName) {
+      payload.name = this.state.queryName;
+    }
+    let promise;
+    if (isEditing) {
+      promise = updateSavedQuery(api, organization.slug, payload);
+    } else {
+      promise = createSavedQuery(api, organization.slug, payload);
+    }
+    promise.then(saved => {
+      const view = EventView.fromSavedQuery(saved);
+      addSuccessMessage(t('Query saved'));
+
+      browserHistory.push({
+        pathname: location.pathname,
+        query: view.generateQueryStringObject(),
+      });
+    });
+  };
+
+  handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const value = event.target.value;
+    this.setState({queryName: value});
+  };
+
+  render() {
+    const {isEditing} = this.props;
+    const buttonText = isEditing ? t('Update') : t('Save');
+
+    return (
+      <DropdownControl
+        alignRight={true}
+        menuWidth="220px"
+        button={({isOpen, getActorProps}) => (
+          <StyledDropdownButton
+            {...getActorProps({isStyled: true})}
+            isOpen={isOpen}
+            showChevron={false}
+          >
+            <StyledInlineSvg src="icon-bookmark" size="14" />
+            {t('Save Search')}
+          </StyledDropdownButton>
+        )}
+      >
+        <SaveQueryContainer onClick={this.swallowEvent}>
+          <StyledInput
+            type="text"
+            placeholder={t('display name')}
+            value={this.state.queryName}
+            onChange={this.handleInputChange}
+          />
+          <Button size="small" onClick={this.handleSave} priority="primary">
+            {buttonText}
+          </Button>
+        </SaveQueryContainer>
+      </DropdownControl>
+    );
+  }
+}
+
+const SaveQueryContainer = styled('li')`
+  padding: ${space(1)};
+`;
+
+const StyledInlineSvg = styled(InlineSvg)`
+  margin-right: 0.33em;
+`;
+
+const StyledInput = styled(Input)`
+  width: 100%;
+  margin-bottom: ${space(1)};
+`;
+
+const StyledDropdownButton = styled(
+  React.forwardRef((prop, ref) => <DropdownButton innerRef={ref} {...prop} />)
```

Comment:
Why do we need to forwardRef here?

Reply:
The forwardRef is needed to preserve refs with `DropdownMenu` which needs the ref to close the dropdown on a click outside the dropdown.

Reply:
`getActorProps` should return an `innerRef` prop